src/internal/modules/invoice-accounts/controllers/invoice-accounts.js

```diff
@@ -53,35 +55,74 @@ const postCompany = async (request, h) => {
 
 const getAddress = async (request, h) => {
   const { regionId, companyId } = request.params;
-  const addresses = await dataService.getCompanyAddresses(companyId);
   // get the session data to check if the address has been set and if it is new or existing
   const session = dataService.sessionManager(request, regionId, companyId);
+  const addresses = await helpers.getAllAddresses(companyId, session);
   // @TODO this might need a mapper to map the session address data to the Company address shape passed to the form
   if (session.address && session.address.addressId === tempId) { addresses.push(session.address); }
   const selectedAddressId = has(session, 'address') ? session.address.addressId : null;
   return h.view('nunjucks/form', {
     ...request.view,
     caption: helpers.getFormTitleCaption(session.viewData.licenceNumber),
     pageTitle: `Select an existing address for ${session.viewData.companyName}`,
-    back: '/manage',
+    back: `/invoice-accounts/create/${regionId}/${companyId}?redirectPath=${session.viewData.redirectPath}`,
     form: sessionForms.get(request, selectAddressForm(request, addresses, selectedAddressId))
   });
 };
 
 const postAddress = async (request, h) => {
   const { regionId, companyId } = request.params;
-  const addresses = await dataService.getCompanyAddresses(companyId);
+  const session = dataService.sessionManager(request, regionId, companyId);
+  const addresses = await helpers.getAllAddresses(companyId, session);
   const schema = selectAddressFormSchema(request.payload);
   const form = forms.handleRequest(selectAddressForm(request, addresses), request, schema);
   if (form.isValid) {
     const { selectedAddress } = forms.getValues(form);
-    dataService.sessionManager(request, regionId, companyId, { address: { addressId: selectedAddress } });
+    if (selectedAddress !== tempId) {
```

Comment:
what if the selected address does have a tempId?

Reply:
`selectedAddress` here is not an address object - rather it's either a GUID or a string that says `new_address`. It will never be null because Joi shouldn't allow it. I hope that makes sense.

I will add a comment to explain this in the code. :)